src/base/TypeUtil.ml

```diff
@@ -65,23 +65,31 @@ module type MakeTEnvFunctor = functor (Q : QualifiedTypes) (R : Rep) -> sig
   module TEnv : sig
     type t
 
+    type restore
+
     (* Make new type environment *)
-    val mk : t
+    val mk : unit -> t
 
     (* Add to type environment *)
-    val addT : t -> R.rep TUIdentifier.t -> TUType.t -> t
+    val addT : t -> R.rep TUIdentifier.t -> TUType.t -> restore list
```

Comment:
Should this not be `val addT : t -> R.rep TUIdentifier.t -> TUType.t -> restore`? 

From an abstract point of view a `restore` is an entirely opaque value, so it seems reasonable that one `restore` should correspond to one operation.